pkg/runtime/serializer/versioning/versioning.go

```diff
@@ -225,51 +137,95 @@ func (c *codec) Decode(data []byte, defaultGVK *unversioned.GroupVersionKind, in
 // Encode ensures the provided object is output in the appropriate group and version, invoking
 // conversion if necessary. Unversioned objects (according to the ObjectTyper) are output as is.
 func (c *codec) Encode(obj runtime.Object, w io.Writer) error {
-	if _, ok := obj.(*runtime.Unknown); ok {
+	switch t := obj.(type) {
+	case *runtime.Unknown:
+		if gv, ok := runtime.PreferredGroupVersion(c.encodeVersion); ok {
+			t.APIVersion = gv.String()
```

Comment:
How does this work? what if the object can't be expressed in this group/version?